### PR TITLE
Add support for the profiles endpoint

### PR DESCRIPTION
**Pull request by surajssd**
* **api: Add endpoint to serve profiles**

Profiles endpoint will allow user to submit information on the way they
want to start the container for the user.

This commit adds support of custom docker image and port selection.

* **frontend: Support profiles endpoint**

Change the landing page to the profiles page and then redirect it to the
terminal page where the container behind starts with user provided
details.

### CODE REVIEW

**Comment by surajssd** [comment]
To try this image with deployment, use the variable `contained_af_image = "surajd/contained.af:2019-06-13-09-55-20"`.

File: docker.go [hunk 1]
```diff
@@ -15,13 +15,37 @@ import (
...
+	return nat.PortSet{
+		port: struct{}{},
+	}, nil
+}
+
 // startContainer starts a docker container and returns the container ID
 // as well as a websocket connection to the attach endpoint.
-func (h *handler) startContainer() (string, *websocket.Conn, error) {
+func (h *handler) startContainer(ctrInfo containerInfo) (string, *websocket.Conn, error) {
+	// pull alpine image if we don't already have it
```
Comment (alban) [inline docker.go]: s/alpine/container/

Comment (surajssd) [reply docker.go]: done

Comment (surajssd) [reply docker.go]: good catch :-)

File: docker.go [hunk 2]
```diff
@@ -15,13 +15,45 @@ import (
...
+	if portStr == "" {
+		return "", nil
+	}
+	port, err := nat.NewPort("tcp", portStr)
+	if err != nil {
+		return "", err
+	}
+	portInt := port.Int()
+	if portInt < 36100 || portInt > 36110 {
+		return "", fmt.Errorf("port not in range [36100, 36110], given: %d", portInt)
```
Comment (alban) [inline docker.go:40]: Maybe a comment saying this range comes from the port range let open by container-escape-bounty:
https://github.com/kinvolk/container-escape-bounty/pull/19

File: docker.go [hunk 3]
```diff
@@ -15,13 +15,45 @@ import (
...
+	}
+	return port, nil
+}
+
 // startContainer starts a docker container and returns the container ID
 // as well as a websocket connection to the attach endpoint.
-func (h *handler) startContainer() (string, *websocket.Conn, error) {
+func (h *handler) startContainer(ctrInfo containerInfo) (string, *websocket.Conn, error) {
+	// Add defaults to the docker image
```
Comment (alban) [inline docker.go]: I don't understand that comment

File: frontend/index.html [hunk 4]
```diff
@@ -7,38 +7,33 @@
...
-        </div>
-
-
-        <script src="js/contained.min.js"></script>
+        <form action="term.html">
+            Image Name: <input type = "text" name = "image" />
+            <br>
+            Contianer Port:
+            <select name="port">
+                <option style="display:none">
```
Comment (alban) [inline frontend/index.html]: Add `/` before the `>`

Comment (alban) [reply frontend/index.html]: What is `display:none` for?

File: frontend/term.html [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+        <meta charset="utf-8">
+        <meta http-equiv="X-UA-Compatible" content="IE=edge">
+        <meta name="viewport" content="width=device-width, initial-scale=1">
+
+        <title>contained.af</title>
```
Comment (alban) [inline frontend/term.html]: We could say "Container Escape Bounty CTF" here too

File: frontend/term.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            <footer class="footer">
+                <div class="row">
+                    <div class="col-md-5">
+                        <p>There is a CTF on every VM instance. <b><em>If</em></b>&nbsp;
+                        you manage to break out of the container, email
+                        <a href="#">to-be-defined@example.com</a>
+                        and you will be rewarded. <br/>
+                        <em>If</em>&nbsp; you bother this email address with anything
+                        that is not the ascii art contents of the flag file you will be
+                        ignored.</p>
```
Comment (alban) [inline frontend/term.html]: The flag is currently not in ascii art :D

https://github.com/kinvolk/container-escape-bounty/blob/b2390f24c52de0f2356a7216e6382f4af28dc325/terraform/flagfiles.tf#L1

File: frontend/term.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                    </div>
+                    <div class="col-md-2"></div>
+                    <div class="col-md-5">
+                        <p>Idea and implementation based on
+                        <a href="https://github.com/genuinetools/contained.af" target="_blank">github.com/genuinetools/contained.af</a>.</p>
```
Comment (alban) [inline frontend/term.html]: maybe you could say:

> initially based on https://github.com/genuinetools/contained.af with modifications on https://github.com/kinvolk/contained.af/tree/container-escape-bounty

File: frontend/term.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                    </div>
+                </div>
+            </footer>
+        </div>
+
+
```
Comment (alban) [inline frontend/term.html]: double empty line

**Review by alban — COMMENTED** [review]
I added some comments. After they are addressed, this can be merged.

File: docker.go (hunk 2 above)
Comment (surajssd) [reply docker.go:40]: done

File: docker.go (hunk 3 above)
Comment (surajssd) [reply docker.go]: updated it

File: frontend/index.html (hunk 4 above)
Comment (surajssd) [reply frontend/index.html]: > What is `display:none` for?

When the first option is this, the dropdown has nothing selected by default, without this the first element in the dropdown is automatically selected. Allowing user to provide blank information.

File: frontend/term.html (hunk 5 above)
Comment (surajssd) [reply frontend/term.html]: done

File: frontend/term.html (hunk 6 above)
Comment (surajssd) [reply frontend/term.html]: removed it

File: frontend/term.html (hunk 7 above)
Comment (surajssd) [reply frontend/term.html]: done

File: frontend/term.html (hunk 8 above)
Comment (surajssd) [reply frontend/term.html]: removed one

**Pull request merged**